magenta-lib/src/main/scala/magenta/tasks/ChangeSetTasks.scala

```diff
@@ -0,0 +1,150 @@
+      case LookupByName(cloudFormationStackName) => CloudFormation.describeStack(cloudFormationStackName, cfnClient)
+      case LookupByTags(tags) => CloudFormation.findStackByTags(tags, reporter, cfnClient)
+    }
+
+    val templateString = templatePath.fetchContentAsString.right.getOrElse(
+      reporter.fail(s"Unable to locate cloudformation template s3://${templatePath.bucket}/${templatePath.key}")
+    )
+
+    val stackName = UpdateCloudFormationTask.nameToCallNewStack(cloudFormationStackLookupStrategy)
+    val template = processTemplate(stackName, templateString, s3Client, stsClient, region, alwaysUploadToS3, reporter)
+
+    val templateParameters = CloudFormation.validateTemplate(template, cfnClient).getParameters.asScala
+      .map(tp => TemplateParameter(tp.getParameterKey, Option(tp.getDefaultValue).isDefined))
+
+    val resolvedAmiParameters: Map[String, String] = amiParameterMap.flatMap { case (name, tags) =>
+      val ami = latestImage(accountNumber)(region.name)(tags)
+      ami.map(name ->)
+    }
+
+    val parameters: Map[String, ParameterValue] =
+      combineParameters(stack, stage, templateParameters, userParameters ++ resolvedAmiParameters)
+
+    reporter.info("Creating Cloudformation change set")
+    reporter.info(s"Stack name: $stackName")
+    reporter.info(s"Change set name: $changeSetName")
+    reporter.info(s"Parameters: $parameters")
+
+    val stackTags = PartialFunction.condOpt(cloudFormationStackLookupStrategy){ case LookupByTags(tags) => tags }
+    val changeSetType = if(maybeCfStack.isEmpty) { ChangeSetType.CREATE } else { ChangeSetType.UPDATE }
+
+    if(changeSetType == ChangeSetType.CREATE && !createStackIfAbsent) {
+      reporter.fail(s"Stack $cloudFormationStackLookupStrategy doesn't exist and createStackIfAbsent is false")
+    } else {
+      CloudFormation.createChangeSet(reporter, changeSetName, changeSetType, stackName, stackTags, template, parameters, cfnClient)
+    }
+  }
+
+  def description = s"Create change set $changeSetName for stack $cloudFormationStackLookupStrategy with ${templatePath.fileName}"
+  def verbose = description
+}
+
+case class CheckChangeSetCreatedTask(
+  region: Region,
+  cloudFormationStackLookupStrategy: CloudFormationStackLookupStrategy,
+  changeSetName: String,
+  duration: Long
+)(implicit val keyRing: KeyRing, artifactClient: AmazonS3) extends Task with RepeatedPollingCheck {
+
+  override def execute(reporter: DeployReporter, stopFlag: => Boolean): Unit = {
+    check(reporter, stopFlag) {
+      val cfnClient = CloudFormation.makeCfnClient(keyRing, region)
+      val stackName = UpdateCloudFormationTask.nameToCallNewStack(cloudFormationStackLookupStrategy)
+
+      val request = new DescribeChangeSetRequest().withChangeSetName(changeSetName).withStackName(stackName)
+      val response = cfnClient.describeChangeSet(request)
+
+      response.getStatus match {
+        case "CREATE_COMPLETE" =>
+          true
+
```

Comment:
Surprising amount of vertical space used for this simple function, are these blank lines needed between the case statements? Very minor point and no worries if it's to be consistent with the rest of the codebase.

Reply:
It was actually for debugging, we've written a test for it so will reformat